Add config for federatedIDP role based authorization.

inthirakumaaran opened the pull request:
### Proposed changes in this pull request

Add config to use federated user's role for authorization. This will enable the federated user to be authorized based on their roles.

Add the following config in the `deployment.toml` to add applications that can use this feature. The default value is none

Eg:
```
[system_applications]
fidp_role_based_authz_enabled_apps=["Console"]
```
### Follow up actions

- Merge the following PR (bump the framework version as well) https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1869/files#diff-2f07f17a325e6ccf395c18b4a7750e38ec4ea093623c8511a14d1fe1b4285fd5



### Checklist (for reviewing)

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

pamodaaw commented:
Don't we need to add this to the identity.xml file as well? Have we removed the support for the old config model?

inthirakumaaran commented:
> Don't we need to add this to the identity.xml file as well? Have we removed the support for the old config model?

Added support to old config model

AnuradhaSK commented on components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java:
```diff
@@ -1707,4 +1712,45 @@ public static List<String> getPropertyAsList(String key) {
+     *
+     * @return Set<String> of applications.
+     */
+
```
remove additional line

pulasthi7 commented on features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2:
```diff
@@ -2902,6 +2902,18 @@
     <FederatedIDPRoleClaimValueAttributeSeparator>{{federated.idp.role_claim_value_attribute_separator}}</FederatedIDPRoleClaimValueAttributeSeparator>
 
+     <!--This enables federated user's role based authorization -->
+     <EnableFederatedIDPRoleBasedAuthorization>{{federated.idp.role_based_authorization}}</EnableFederatedIDPRoleBasedAuthorization>
```
the config name need to be revisited.
[federated] as top level config doesn't make sense

if this will be used only for system apps, better put this under [system_applications]

inthirakumaaran replied:
If we generalize this solution for product-is, we can use this flow for all the applications (not bounded to system applications) & it will be more like improvement to federated login flow (regardless of application type), WDYT?

inthirakumaaran replied:
Going with `system_applications` top level

inthirakumaaran commented on features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml:
```diff
@@ -2131,6 +2131,13 @@
     <FederatedIDPRoleClaimValueAttributeSeparator>,</FederatedIDPRoleClaimValueAttributeSeparator>
 
+    <!--This enables federated user's role based authorization -->
+    <EnableFederatedIDPRoleBasedAuthorization>false</EnableFederatedIDPRoleBasedAuthorization>
```
Merge these two configs & if at least one application exist consider as feature is enabled

inthirakumaaran commented on features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/org.wso2.carbon.identity.core.server.feature.default.json:
```diff
@@ -611,6 +611,7 @@
   "authentication.adaptive.execution_supervisor.timeout": "500ms",
 
   "federated.idp.role_claim_value_attribute_separator": ",",
+  "federated.idp.role_based_authorization": false,
```
Handle it from the implementation

inthirakumaaran commented on features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml:
```diff
@@ -2131,6 +2131,13 @@
+    <EnableFederatedIDPRoleBasedAuthorization>false</EnableFederatedIDPRoleBasedAuthorization>
+    <!-- Following applications can use federated user's role based authorization -->
+    <FederatedIDPRoleBasedAuthorizationApplications>
+        <ApplicationName></ApplicationName>
```
```suggestion
        <AppliName></ApplicationName>
```

pamodaaw commented on features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml:
```diff
@@ -2149,6 +2149,11 @@
 
+    <!-- Following applications can use federated user role based authorization -->
+    <FIdPRoleBasedAuthzApplications>
+        <AppName></AppName>
```
does it work as expected with the empty config?

inthirakumaaran replied:
yes, we check the list for SP name & having an empty list will also works here

mefarazath replied:
Since this is a sample value, can we have a commented XML tag with a sample app name

pulasthi7 commented on features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2:
```diff
@@ -2923,6 +2923,15 @@
+     {% if system_applications.fidp.role_based_authz_apps is defined %}
+     <!-- Following applications can use federated user role based authorization -->
+     <FIdPRoleBasedAuthzApplications>
+         {% for application in system_applications.fidp.role_based_authz_apps %}
```
```suggestion
         {% for application in system_applications.fidp_role_based_authz_enabled_apps %}
```
